hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactSerializationTest.java

```diff
@@ -430,6 +433,120 @@ public void testReflectiveSerializer_withFieldsWithOtherSerializationMechanisms_
         assertTrue(data.isCompact());
     }
 
+    @Test
+    public void testCompactWriterThrowsExceptionWhileWritingWhenSchemaDoesNotHaveThatField() {
+        FooBarSerializer serializer = spy(new FooBarSerializer());
+        FooBar fooBar = new FooBar("foo", 42L);
+
+        SchemaService schemaService = CompactTestUtil.createInMemorySchemaService();
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
+        compactSerializationConfig.addSerializer(serializer);
+        SerializationService serializationService = new DefaultSerializationServiceBuilder()
+                .setSchemaService(schemaService)
+                .setConfig(new SerializationConfig().setCompactSerializationConfig(compactSerializationConfig))
+                .build();
+
+        // Add correct schema to classToSchemaMap
+        serializationService.toData(fooBar);
+        // Change serializer's write method
+        doAnswer(invocation -> {
+            CompactWriter writer = invocation.getArgument(0);
+            FooBar value = invocation.getArgument(1);
+            writer.writeString("foo", value.getFoo());
+            writer.writeInt64("bar", value.getBar());
+            writer.writeString("baz", "a");
+            return null;
+        }).when(serializer).write(any(), any());
+        // This should throw because schema does not have a baz field.
+        assertThatThrownBy(() -> {
+            serializationService.toData(fooBar);
+        }).isInstanceOf(HazelcastSerializationException.class)
+          .hasCauseInstanceOf(HazelcastSerializationException.class)
+          .hasStackTraceContaining("Invalid field name");
+    }
+
+    @Test
+    public void testCompactWriterThrowsExceptionWhileWritingWhenFieldTypeDoesNotMatch() {
+        FooBar fooBar = new FooBar("foo", 42L);
+        FooBarSerializer serializer = spy(new FooBarSerializer());
+
+        SchemaWriter schemaWriter = new SchemaWriter(serializer.getTypeName());
+        serializer.write(schemaWriter, fooBar);
+        Schema schema = schemaWriter.build();
+
+        SchemaService schemaService = CompactTestUtil.createInMemorySchemaService();
+        schemaService.put(schema);
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
+        compactSerializationConfig.addSerializer(serializer);
+        SerializationService serializationService = new DefaultSerializationServiceBuilder()
+                .setSchemaService(schemaService)
+                .setConfig(new SerializationConfig().setCompactSerializationConfig(compactSerializationConfig))
+                .build();
```

Comment:
```suggestion
        SerializationService serializationService = createSerializationService(() -> serializer);
```

this one can also be simplified